### DISCUSSION TITLE
Supporting remix's new optional route segments

### DISCUSSION THREAD

**Discussion by aniravi24** [sandulat/routes-gen#27] [Ideas]
See the following release notes for the description on remix's new optional route segments feature:
https://github.com/remix-run/remix/releases/tag/remix%401.9.0

I thought it would be worth starting a discussion on what types to generate.

Using the example from the release notes:

The path `/(one)/($two)/(three).($four)` matches

```
/
/one
/one/param1
/one/param1/three
/one/param1/three/param2
```

The options that came to mind as to what we can generate:

Option 1: Generate each route strictly and don't allow undefined paths

```
{
  "/": Record<string, never>;
  "/one": Record<string, never>;
  "/one/:param1": {param1: string};
  "/one/:param1/three": {param1: string};
  "/one/:param1/three/:param2": {param2: string};
}
````

Option 2: Generate just the minimum number of routes with optional params
```
{
  "/": Record<string, never>;
  "/one/:param1?": {param1?: string};
  "/one/:param1/three/:param2?": {param1?: string; param2: string};
}
```

Option 3: Generate both optional param and non-optional routes
```
{
  "/": Record<string, never>;
  "/one": Record<string, never>;
  "/one/:param1": {param1: string};
  "/one/:param1?": {param1?: string};
  "/one/:param1/three": {param1: string};
  "/one/:param1/three/:param2": {param2: string};
  "/one/:param1/three/:param2?": {param2?: string};
}
```

Thoughts? I think option 1 or option 3 seem most interesting. As a user of this library, I'm not sure I would like option 2 because I'd always have to specify the optional param in my route even if I'm not using it.

**Comment by brandonpittman** [comment]
I guess option 3 would be the best. Really looking forward to this getting added. Started using optional routes today and now TypeScript is so very mad at me. 😄

**Comment by sandulat** [comment]
This is now released under `routes-gen@0.6.0` (https://github.com/sandulat/routes-gen/pull/32). Thank you so much @brandonpittman!

@aniravi24 
Second parameter is optional when there are only optional params:
`route('/articles/:lang?')` => `/articles`
`route('/articles/:lang?', {})` => `/articles`
`route('/articles/:lang?', { lang: 'en' })` => `/articles/en`

Second parameter is required when there is at least one required param:
`route('/articles/:lang?/:articleId', { articleId: '1' })` => `/articles/1`
`route('/articles/:lang?/:articleId', { articleId: '1', lang: 'en'  })` => `/articles/en/1`

**Reply by aniravi24** [reply]
Awesome, thanks!
